File: Python/자료구조_알고리즘/재귀/하노이의탑.py
# 하노이의 탑 자료구조 공부 10월14일

# 재귀

def hanoi(n,A,B,C):
  if n == 1:
    print(f"1번 타워를 {A}에서 {C}로 이동")
  else:
    hanoi(n-1,A,C,B)
    print(f"{n}번타워를 {A}에서 {C}로 이동")
    hanoi(n-1,B,A,C)

n = 4
print(hanoi(n,'A','B','C'))

File: JavaScript/2주차/재귀/하노이의탑.js
let fs = require('fs');
let input = fs.readFileSync('io.txt').toString().trim();
let num = Number(input);

// 이동 횟수 계산 (2^num - 1)
let answer = 2n ** BigInt(num) - 1n; // N이 100까지 가능하므로 BigInt 사용
console.log(answer.toString());

// N이 20보다 크면 과정 생략
if (num > 20) {
  process.exit(0);
}

let result = [];

function hanoi(n, from, target, temp) {
  if (n === 1) {
    result.push(`${n}번 원판을 ${from}에서 ${target}로 이동`);
    return;
  }
  hanoi(n - 1, from, temp, target);
  result.push(`${n}번 원판을 ${from}에서 ${target}로 이동`);
  hanoi(n - 1, temp, target, from);
}

hanoi(num, 'A', 'C', 'B');
console.log(result.join('\n'));
